src/main/java/net/minecraftforge/server/permission/nodes/PermissionNode.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Minecraft Forge
+ * Copyright (c) 2016-2021.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.server.permission.nodes;
+
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.server.level.ServerPlayer;
+
+/**
+ * @see PermissionNode#PermissionNode(String, PermissionType, PermissionResolver, PermissionDynamic[])
+ * @param <T> PermissionType see {@link PermissionTypes}
+ */
+public class PermissionNode<T>
+{
+   private final String nodeName;
+   private final PermissionType<T> type;
+   private final PermissionResolver<T> resolver;
+   private final PermissionDynamic[] dynamics;
```

Comment:
`PermissionDynamic<?>`

Also, why an array and not a `List` or `Collection`?

Reply:
In favor of being able to use varargs. I know that you have fancy alternatives for this in Kotlin.
But in Java this is the cleanest solution.